File: src/docs/website/site/chromium-os/developer-library/guides/testing/web-testing/iwa/test-iwa-interaction.py
# Copyright 2026 The Chromium Authors
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
"""Selenium Test Script for Kitchen Sink IWA Interaction.

This script connects to a remote ChromeOS instance, installs the
Kitchen Sink Isolated Web App (IWA) from its web bundle, then
launches it, and interacts with its UI to simulate sending and
receiving messages via direct sockets.

Prerequisites:
1. A CrOS device or VM setup for remote testing.
2. Chrome on the device must be launched with the following flags in
   /etc/chrome_dev.conf:
   - `--remote-debugging-port=<port>`
   - `--enable-devtools-pwa-handler`
   - `--enable-features=IsolatedWebAppDevMode`
"""

import argparse
import logging
import sys
import time

from selenium import webdriver
from selenium.common import exceptions
from selenium.webdriver.chrome import options
from selenium.webdriver.chrome import service
from selenium.webdriver.common import by
from selenium.webdriver.remote import webelement
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support import ui

# Configure logging.
logging.basicConfig(level=logging.INFO, format="[%(levelname)s] %(message)s")

# IWA specific constants.
_WEB_BUNDLE_ID = "aiv4bxauvcu3zvbu6r5yynoh4atkzqqaoeof5mwz54b4zfywcrjuoaacai"
_IWA_VERSION = "0.18.0"
_WEB_BUNDLE_URL = ("https://github.com/chromeos/iwa-sink/releases/download/"
                   f"v{_IWA_VERSION}/iwa-sink.swbn")
_MANIFEST_ID = f"isolated-app://{_WEB_BUNDLE_ID}/"

_MSG_LEFT = "Sing me a song!"
_MSG_RIGHT = "Hello from the right side!"


def get_shadow_root(driver: webdriver.Chrome,
                    host_element: webelement.WebElement) -> object:
    """Retrieves the shadow root of a given web element."""
    return driver.execute_script("return arguments[0].shadowRoot", host_element)


def run_iwa_interaction_test(driver_path: str, port: int) -> bool:
    """Installs and tests the Kitchen Sink IWA.

    Args:
        driver_path: Path to the chromedriver executable.
        port: The remote-debugging-port flag value on the DUT.

    Returns:
        True on success, False on error.
    """
    debugger_addr = f"127.0.0.1:{port}"
    logging.info("Connecting to browser at %s...", debugger_addr)

    chrome_opts = options.Options()
    chrome_opts.add_experimental_option("debuggerAddress", debugger_addr)
    chrome_svc = service.Service(executable_path=driver_path)

    try:
        with webdriver.Chrome(service=chrome_svc,
                              options=chrome_opts) as driver:
            logging.info("Successfully connected to the browser.")
            original_window = driver.current_window_handle

            # Install IWA.
            logging.info("Installing IWA from URL: %s", _WEB_BUNDLE_URL)
            driver.execute_cdp_cmd(
                "PWA.install",
                {
                    "manifestId": _MANIFEST_ID,
                    "installUrlOrBundleUrl": _WEB_BUNDLE_URL,
                },
            )
            # Allow time for installation to propagate.
            time.sleep(5)

            # Launch IWA.
            logging.info("Launching IWA with manifest ID: %s", _MANIFEST_ID)
            driver.execute_cdp_cmd("PWA.launch", {"manifestId": _MANIFEST_ID})

            # Wait for window switch.
            wait = ui.WebDriverWait(driver, 15)
            wait.until(lambda d: len(d.window_handles) > 1)

            # Switch to IWA window.
            iwa_window_handle = [
                h for h in driver.window_handles if h != original_window
            ][0]
            driver.switch_to.window(iwa_window_handle)
            logging.info("Switched to IWA window: '%s'", driver.title)

            # --- UI Interaction ---
            wait_long = ui.WebDriverWait(driver, 30)

            logging.info("Waiting for 'Create new socket connection' button...")
            add_socket_button = wait_long.until(
                expected_conditions.presence_of_element_located(
                    (by.By.ID, "addSocketButton")))
            add_socket_button.click()
            logging.info("Clicked 'Create new socket connection' button.")

            # Interact with Shadow DOM.
            left_comp = wait_long.until(
                expected_conditions.presence_of_element_located(
                    (by.By.TAG_NAME, "socket-connection")))
            right_comp = wait_long.until(
                expected_conditions.presence_of_element_located(
                    (by.By.TAG_NAME, "socket-server")))

            left_root = get_shadow_root(driver, left_comp)
            right_root = get_shadow_root(driver, right_comp)

            # Message: Left -> Right.
            logging.info("Sending message from Left to Right...")
            left_input = left_root.find_element(by.By.CSS_SELECTOR,
                                                "input#messageInput")
            left_send_btn = left_root.find_element(by.By.CSS_SELECTOR,
                                                   "button#sendButton")

            left_input.send_keys(_MSG_LEFT)
            left_send_btn.click()

            wait.until(lambda d: _MSG_LEFT in right_root.find_element(
                by.By.CSS_SELECTOR, "#log").text)
            logging.info("Verified '%s' appeared on the right side.", _MSG_LEFT)

            # Message: Right -> Left.
            logging.info("Sending message from Right to Left...")
            right_input = right_root.find_element(by.By.CSS_SELECTOR,
                                                  "input#messageInput")
            right_send_btn = right_root.find_element(by.By.CSS_SELECTOR,
                                                     "button#sendButton")

            right_input.send_keys(_MSG_RIGHT)
            right_send_btn.click()

            wait.until(lambda d: _MSG_RIGHT in left_root.find_element(
                by.By.CSS_SELECTOR, "#log").text)
            logging.info("Verified '%s' appeared on the left side.", _MSG_RIGHT)

            # Cleanup within the context.
            try:
                logging.info("Uninstalling IWA: %s", _MANIFEST_ID)
                driver.execute_cdp_cmd("PWA.uninstall",
                                       {"manifestId": _MANIFEST_ID})
            except exceptions.WebDriverException as e:
                logging.warning("Could not uninstall IWA during cleanup: %s", e)

            logging.info("IWA test successful.")
            return True

    except exceptions.WebDriverException as e:
        logging.error("An error occurred during IWA interaction test: %s",
                      e,
                      exc_info=True)
        return False


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--driver-path", default="./chromedriver")
    parser.add_argument("--port", type=int, default=9222)
    opts = parser.parse_args(argv)

    return 0 if run_iwa_interaction_test(opts.driver_path, opts.port) else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: src/docs/website/site/chromium-os/developer-library/guides/testing/web-testing/iwa/test-iwa-interaction.ts
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @fileoverview Playwright Test Script for Kitchen Sink IWA Interaction
 * (TypeScript).
 *
 * This script connects to a remote ChromeOS instance, installs the
 * Kitchen Sink Isolated Web App (IWA) from its web bundle, then
 * launches it, and interacts with its UI to simulate sending and
 * receiving messages via direct sockets.
 *
 * Prerequisites:
 * 1. A CrOS device or VM setup for remote testing.
 * 2. Chrome must be launched with:
 * - `--remote-debugging-port=<port>`
 * - `--enable-devtools-pwa-handler`
 * - `--enable-features=IsolatedWebAppDevMode`
 */

import * as fs from 'fs';
import * as path from 'path';
import {Browser, BrowserContext, CDPSession, chromium, Page} from 'playwright';

// Constants
const KITCHEN_SINK_IWA_WEB_BUNDLE_ID =
    'aiv4bxauvcu3zvbu6r5yynoh4atkzqqaoeof5mwz54b4zfywcrjuoaacai';
const IWA_MANIFEST_ID = `isolated-app://${KITCHEN_SINK_IWA_WEB_BUNDLE_ID}/`;
const KITCHEN_SINK_IWA_VERSION = '0.18.0';
const KITCHEN_SINK_IWA_WEB_BUNDLE_URL =
    `https://github.com/chromeos/iwa-sink/releases/download/v${
        KITCHEN_SINK_IWA_VERSION}/iwa-sink.swbn`;

const MESSAGE_FROM_LEFT = 'Sing me a song!';
const MESSAGE_FROM_RIGHT = 'Hello from the right side!';

// Color constants for console output.
const COLORS = {
  normal: '\x1b[m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
};

// Helper for colorless environments.
if (process.stderr && !process.stderr.hasColors()) {
  (Object.keys(COLORS) as (keyof typeof COLORS)[]).forEach(
      (key) => COLORS[key] = '');
}

/**
 * Helper to send CDP commands with logging.
 */
async function sendCDPCommand(
    session: CDPSession, command: string, params: object): Promise<any> {
  console.log(COLORS.green, `Sending CDP command: ${command}`, COLORS.normal);
  try {
    return await session.send(command as any, params as any);
  } catch (e: any) {
    console.error(
        COLORS.red, `CDP Error (${command}): ${e.message}`, COLORS.normal);
    throw e;
  }
}

async function runIWAInteractionTest() {
  const port = process.argv[2] || '9222';
  const browserURL = `http://localhost:${port}`;

  let browser: Browser|null = null;
  let context: BrowserContext|null = null;
  let managementPage: Page|null = null;
  let iwaPage: Page|null = null;

  try {
    console.log(`Connecting to browser at ${browserURL}...`);
    browser = await chromium.connectOverCDP(browserURL);
    console.log('Successfully connected.');

    // Get default context or create one.
    if (browser.contexts().length > 0) {
      context = browser.contexts()[0];
    } else {
      context = await browser.newContext();
    }

    // Create a session on a management page to perform PWA commands.
    managementPage = await context.newPage();
    const client = await context.newCDPSession(managementPage);

    // 1. Install IWA.
    console.log('Installing IWA...');
    await sendCDPCommand(client, 'PWA.install', {
      manifestId: IWA_MANIFEST_ID,
      installUrlOrBundleUrl: KITCHEN_SINK_IWA_WEB_BUNDLE_URL,
    });

    // 2. Launch IWA.
    console.log('Launching IWA...');
    await sendCDPCommand(client, 'PWA.launch', {manifestId: IWA_MANIFEST_ID});

    // 3. Wait for IWA Window.
    console.log('Waiting for IWA window to appear...');
    iwaPage = await context.waitForEvent('page', {
      predicate: (page) => page.url().startsWith('isolated-app://'),
      timeout: 15000
    });

    await iwaPage.bringToFront();
    console.log('IWA Launched. Title:', await iwaPage.title());

    // 4. UI Interaction.
    console.log('Interacting with IWA UI...');

    // Click 'Create new socket connection'.
    await iwaPage.click('#addSocketButton');

    // Define Locators for Shadow DOM components.
    // Playwright Locators pierce Shadow DOM boundaries automatically.
    const leftSocket = iwaPage.locator('socket-connection');
    const rightSocket = iwaPage.locator('socket-server');

    // Wait for components to be ready.
    await leftSocket.waitFor();
    await rightSocket.waitFor();
    console.log('Socket components found.');

    // --- TEST 1: Left -> Right ---
    console.log(
        `\n--- Test 1: Sending "${MESSAGE_FROM_LEFT}" from Left to Right ---`);

    const leftInput = leftSocket.locator('input#messageInput');
    const leftSendBtn = leftSocket.locator('button#sendButton');

    await leftInput.fill(MESSAGE_FROM_LEFT);
    await leftSendBtn.click();

    // Verify on Right Side.
    // Use strict selector 'p#log' to avoid ambiguity with <socket-log>.
    const rightLog =
        rightSocket.locator('p#log', {hasText: MESSAGE_FROM_LEFT});
    await rightLog.waitFor({timeout: 10000});
    console.log(
        COLORS.green, 'Verified message received on Right side.',
        COLORS.normal);

    // --- TEST 2: Right -> Left ---
    console.log(
        `\n--- Test 2: Sending "${MESSAGE_FROM_RIGHT}" from Right to Left ---`);

    const rightInput = rightSocket.locator('input#messageInput');
    const rightSendBtn = rightSocket.locator('button#sendButton');

    await rightInput.fill(MESSAGE_FROM_RIGHT);
    await rightSendBtn.click();

    // Verify on Left Side.
    const leftLog = leftSocket.locator('p#log', {hasText: MESSAGE_FROM_RIGHT});
    await leftLog.waitFor({timeout: 10000});
    console.log(
        COLORS.green, 'Verified message received on Left side.', COLORS.normal);

    console.log('IWA interaction test passed.');

  } catch (error: any) {
    console.error(
        COLORS.red, '--- An error occurred during the test ---', COLORS.normal);
    console.error(error);

    // Screenshot on failure.
    if (iwaPage) {
      try {
        const screenshotDir = path.join(__dirname, 'screenshots');
        if (!fs.existsSync(screenshotDir)) fs.mkdirSync(screenshotDir);
        const screenshotPath =
            path.join(screenshotDir, 'failed_iwa_test_ts.png');
        await iwaPage.screenshot({path: screenshotPath});
        console.log(
            COLORS.yellow, `Screenshot saved to ${screenshotPath}`,
            COLORS.normal);
      } catch (e) {
        console.error('Failed to take screenshot:', e);
      }
    }
    process.exitCode = 1;
  } finally {
    console.log('\nCleanup Phase...');
    if (browser && context) {
      try {
        // Need a valid page session to uninstall. Ensure managementPage is open
        // or create new one.
        let cleanupSessionPage = managementPage;
        if (!cleanupSessionPage || cleanupSessionPage.isClosed()) {
          cleanupSessionPage = await context.newPage();
        }
        const client = await context.newCDPSession(cleanupSessionPage);

        console.log(`Uninstalling IWA: ${IWA_MANIFEST_ID}`);
        await sendCDPCommand(
            client, 'PWA.uninstall', {manifestId: IWA_MANIFEST_ID});
        console.log('IWA uninstalled.');

        if (cleanupSessionPage !== managementPage) {
          await cleanupSessionPage.close();
        }
      } catch (e: any) {
        console.warn(
            COLORS.yellow,
            'Warning: Failed to uninstall IWA during cleanup:', e.message,
            COLORS.normal);
      }

      if (managementPage && !managementPage.isClosed()) {
        await managementPage.close();
      }
      await browser.close();
    }
    console.log('Test finished.');
  }
}

runIWAInteractionTest();
